### PR TITLE
Search filter template component

### PR DESCRIPTION
**Pull request by dhlevi**
Adding common componets for handling searching, paging, sorting.

This is a big one, so adding everyone for a review. There may be a few changes and updates, but its worth getting an eye or two on it early due to the volume of changes.

I've added a search component and handlers to the existing table component. The reason for this was because we may create tables without a search, but we only ever create searches with a table template, so it made sense to keep them together (but there is a case included for showing a search without a table template). This also allowed me to share resources.

app-table-template has been modified to include a number of new inputs:

```
<app-table-template [columns]="documentTableColumns"
                    [data]="documentTableData"
                    (onColumnSort)='setColumnSort($event)'
                    (onSelectedRow)='updateSelectedRow($event)'
                    (onPageNumUpdate)='getPaginatedDocs($event)'
                    [showMoreLoader]="true"
                    [hidePager]="true"
                    [showTableTemplate]="true"
                    [showSearch]="true"
                    (onSearch)='executeSearch($event)'
                    [showAdvancedSearch]="true"
                    [filters]="filters"
                    [persistenceId]="'globalProjSearch'"
                    [searchDisclaimer]="constants.searchDisclaimer"
                    [showSearchHelp]="false"
                    [buttonFilter]="buttonFilterType">
</app-table-template>
```

None of these will affect any existing table in the application. Default values are set on all new inputs. All tables will function as normal, even with these changes. However, if you want to use the shared component:

set showSearch to true, and showAdvancedSearch to true for adding filters. Create and wire in a "filters" array of FilterObjects. A persistenceId will automatically persist the table filters so navigating back will auto-fill the original settings. Search disclaimer is fairly obvious, as is showSearchHelp. If you want to use the search component without a table, set showTableTemplate to false. This will create a stub TableData object for handling any paging/sorting needed. The showMore/hidePager functionality will also be unaffected if you have a table but are disabling paging in favour of using that query option.

In the code, all you need to do to dynamically generate search filters is create a FilterObject and push it onto the filters array

```
milestoneFilter = new FilterObject('milestone', 'Milestone', null, [], [], { name: 'legislation', labelPrefix: null, labelPostfix: ' Act Terms' });
```

FilterObject accepts and id, title, date filter settings, option array, selected options aray, and a group settings object. Take a look at the code for more examples. I've updated Project, Document and Project Document search to use the new template. It reduces the code considerably.

To add a filter collection (the "More Filters..." button), add an array of filters to an existing FilterObject.

```
      let filterCollection = new FilterObject('moreFilters', 'More Filters...');
      filterCollection.collection = [this.proponentFilter,
                                     this.regionFilter,
                                     this.iaacFilter,
                                     this.phaseFilter];
```

The 'buttonFilter' option accepts a single FilterObject defined as above (without a date filter, they will be ignored) and will generate a row of buttons to use as the filter, in place of the dropdown box. This will display with advanced search turned off as well. See Admin side activity post table for an example.

The component includes an event emitter when clicking the "search" button, so the actual functionality of the search can be handled by the parent component. The component returns a small package containing keywords and API filters. From there, it's up to the component to handle conversion to filterForUI/URL and persistence as needed.

```
let searchPackage = {
      filterForAPI: newFilters,
      keywords: this.keywords
}
```

I've added methods in the upgraded pages to show how to convert the filterorAPI object into the UI/URL objects, so we can completely gut out all of the filterToParam, paramToFilter stuff.

Search is all done in the page without firing a submit, so the resolvers are no longer needed unless you're pre-loading data and code tables from the DB.

See ticket [EE-972](https://bcmines.atlassian.net/browse/EE-972)

### CODE REVIEW

File: src/app/projects/project-list/project-list.component.ts [hunk 1]
```diff
@@ -487,47 +287,55 @@ export class ProjectListComponent implements OnInit, OnDestroy {
...
+    // build filterForUI/URL from the new filterForAPI object
+    this.createFilterForURL();
 
+    this.getPaginatedProjects(this.tableParams.currentPage);
+  }
+
+  createFilterForURL() {
+    this.filterForURL = {};
+    // for each key in filterForAPI
+    Object.keys(this.filterForAPI).forEach(key => {
```
Comment (KitArmstrong) [inline src/app/projects/project-list/project-list.component.ts:304]: Does looping need to occur? I looks like this is just copying the object. Could it be `this.filterForUrl = {...this.filterForApi}`

File: src/app/search/search.component.ts [hunk 2]
```diff
@@ -395,41 +235,44 @@ collectionFilterToParams(params, name, identifyBy) {
...
+    this.createFilterForURL();
 
-    this.setParamsFromFilters(params);
-    this.setFiltersFromParams(params);
+    this.getPaginated(this.tableParams.currentPage);
+  }
+
+  createFilterForURL() {
+    // for each key in filterForAPI
+    Object.keys(this.filterForAPI).forEach(key => {
```
Comment (KitArmstrong) [inline src/app/search/search.component.ts:250]: Same comment as above.

File: src/app/shared/components/table-template/filter-object.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
+/**
```
Comment (KitArmstrong) [inline src/app/shared/components/table-template/filter-object.ts:1]: Great documentation.

**Comment by dhlevi** [comment]
URL params are not read by the component to auto-populate. That's left as a task for the parent component if it wants to handle URL parsing and pre-set the filterFor... objects. That said, it's pretty trivial to add the functionality so I can move this into the component init to preset URL passed-in values.

**Comment by dhlevi** [comment]
> Code looks good, new filterObject should make filters easier to handle across our searches. Couple minor behaviour issues:
> 
> * page size update when on page 2, results in empty table with just table header displayed
> * odd behaviour when changing page size and search terms. Page size is either the last set page size or the search returns a blank table
> * changing search terms/filters the new search doesn't go back to page 1 if you were on a different page

Paging/changing page size is likely not emitting the "search" event to the parent (hence the empty table), or the parent is overriding the handler. I'll check the updated sections and see as it's probably specific to certain tables handling of the events

File: src/app/projects/project-list/project-list.component.ts (hunk 1 above)
Comment (dhlevi) [reply src/app/projects/project-list/project-list.component.ts:304]: Yep, for sure. Originally I assumed filterForAPI and filterForURL were different (as they exist in two different objects in the original code). Turns out... they're not at all. You can use the objects interchangably. So the right answer here is not so much copying, but just replacing filterForURL to use filterForAPI I think

Comment (dhlevi) [reply src/app/projects/project-list/project-list.component.ts]: Fixed

File: src/app/search/search.component.ts (hunk 2 above)
Comment (dhlevi) [reply src/app/search/search.component.ts]: Fixed

**Comment by dhlevi** [comment]
I've updated some handling on paging/searching for fixing active page selection issues when you filter or resize that should correct the issues about the page not visually resetting to 1

**Pull request merged**